File: generated_code/GeminiPro/Java/070_SplayTree.java
/**
 * A self-contained Java class implementing a Splay Tree for an integer set.
 * Splay trees are self-balancing binary search trees with the additional property
 * that recently accessed elements are quick to access again.
 */
public class SplayTree {

    /**
     * Represents a node in the Splay Tree.
     */
    private class Node {
        int key;
        Node left, right;

        /**
         * Constructs a new node with the given key.
         * @param key The integer key for the node.
         */
        Node(int key) {
            this.key = key;
            this.left = null;
            this.right = null;
        }
    }

    private Node root;

    /**
     * Constructs an empty Splay Tree.
     */
    public SplayTree() {
        this.root = null;
    }

    /**
     * Performs a right rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node rightRotate(Node x) {
        Node y = x.left;
        x.left = y.right;
        y.right = x;
        return y;
    }

    /**
     * Performs a left rotation on the subtree rooted at x.
     * @param x The root of the subtree to rotate.
     * @return The new root of the subtree after rotation.
     */
    private Node leftRotate(Node x) {
        Node y = x.right;
        x.right = y.left;
        y.left = x;
        return y;
    }

    /**
     * The core splaying operation. It brings the node with the given key
     * (or the last accessed node if the key is not found) to the root of the tree.
     * This is a recursive, bottom-up splaying implementation.
     *
     * @param node The root of the current subtree.
     * @param key The key to splay for.
     * @return The new root of the splayed subtree.
     */
    private Node splay(Node node, int key) {
        // Base cases: node is null or key is present at root
        if (node == null || node.key == key) {
            return node;
        }

        // Key lies in the left subtree
        if (key < node.key) {
            // Key is not in the tree, we are done
            if (node.left == null) {
                return node;
            }

            // Zig-Zig (Left Left)
            if (key < node.left.key) {
                node.left.left = splay(node.left.left, key);
                node = rightRotate(node);
            }
            // Zig-Zag (Left Right)
            else if (key > node.left.key) {
                node.left.right = splay(node.left.right, key);
                if (node.left.right != null) {
                    node.left = leftRotate(node.left);
                }
            }
            // Perform the second rotation for Zig-Zig or the only rotation for Zig
            return (node.left == null) ? node : rightRotate(node);
        }
        // Key lies in the right subtree
        else { // key > node.key
            // Key is not in tree, we are done
            if (node.right == null) {
                return node;
            }

            // Zig-Zag (Right Left)
            if (key < node.right.key) {
                node.right.left = splay(node.right.left, key);
                if (node.right.left != null) {
                    node.right = rightRotate(node.right);
                }
            }
            // Zig-Zig (Right Right)
            else if (key > node.right.key) {
                node.right.right = splay(node.right.right, key);
                node = leftRotate(node);
            }
            // Perform the second rotation for Zig-Zig or the only rotation for Zig
            return (node.right == null) ? node : leftRotate(node);
        }
    }

    /**
     * Inserts a key into the splay tree. If the key already exists, the tree remains unchanged.
     * After insertion, the new node becomes the root of the tree.
     * @param key The integer key to insert.
     */
    public void insert(int key) {
        // If the tree is empty, create a new root
        if (root == null) {
            root = new Node(key);
            return;
        }

        // Splay the tree for the key. This brings the closest node to the root.
        root = splay(root, key);

        // If the key is already present, do nothing
        if (root.key == key) {
            return;
        }

        Node newNode = new Node(key);

        // If the new key is smaller than the root's key
        if (key < root.key) {
            newNode.right = root;
            newNode.left = root.left;
            root.left = null;
        }
        // If the new key is larger than the root's key
        else {
            newNode.left = root;
            newNode.right = root.right;
            root.right = null;
        }
        root = newNode; // The new node is the new root
    }

    /**
     * Searches for a key in the splay tree.
     * If the key is found, it is splayed to the root.
     * If the key is not found, the last accessed node on the search path is splayed to the root.
     * @param key The integer key to search for.
     * @return The key if found, otherwise null.
     */
    public Integer search(int key) {
        if (root == null) {
            return null;
        }
        root = splay(root, key);
        if (root.key == key) {
            return root.key; // Autoboxing to Integer
        }
        return null;
    }

    /**
     * Deletes a key from the splay tree.
     * If the key is present, it is first splayed to the root and then removed.
     * The new root will be the largest element in the left subtree.
     * @param key The integer key to delete.
     */
    public void delete(int key) {
        // If the tree is empty, there is nothing to delete
        if (root == null) {
            return;
        }

        // Splay the tree for the key
        root = splay(root, key);

        // If the key is not in the tree, do nothing
        if (root.key != key) {
            return;
        }

        // Now, the node to be deleted is at the root.
        // If there is no left subtree, the right subtree becomes the new tree.
        if (root.left == null) {
            root = root.right;
        } else {
            // Save the right subtree
            Node rightSubtree = root.right;
            // The new root is the root of the left subtree
            root = root.left;
            // Splay the left subtree to bring its largest element (predecessor of the deleted key) to its root.
            // Splaying for the original key achieves this.
            root = splay(root, key);
            // Attach the original right subtree to the new root.
            // Since the new root is the max element of the left subtree, it has no right child.
            root.right = rightSubtree;
        }
    }
}

File: generated_code/GeminiPro/Python/070_SplayTree.py
class SplayTree:
    """
    A complete, self-contained implementation of a Splay Tree.

    This class implements a dictionary-like set for storing unique integer keys.
    It supports insert, delete, and search operations. The key feature is the
    splaying operation, which moves an accessed node to the root of the tree,
    optimizing for future accesses to the same node.
    """

    class Node:
        """A node in the splay tree."""
        def __init__(self, key):
            self.key = key
            self.parent = None
            self.left = None
            self.right = None

    def __init__(self):
        """Initializes an empty Splay Tree."""
        self.root = None

    def _left_rotate(self, x):
        """Performs a left rotation on node x."""
        y = x.right
        if y:
            x.right = y.left
            if y.left:
                y.left.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.left:
            x.parent.left = y
        else:
            x.parent.right = y
        if y:
            y.left = x
        x.parent = y

    def _right_rotate(self, x):
        """Performs a right rotation on node x."""
        y = x.left
        if y:
            x.left = y.right
            if y.right:
                y.right.parent = x
            y.parent = x.parent
        if not x.parent:
            self.root = y
        elif x == x.parent.right:
            x.parent.right = y
        else:
            x.parent.left = y
        if y:
            y.right = x
        x.parent = y

    def _splay(self, x):
        """
        Performs the splay operation on node x, moving it to the root.
        """
        while x.parent:
            parent = x.parent
            grandparent = parent.parent
            if not grandparent:  # Zig case
                if x == parent.left:
                    self._right_rotate(parent)
                else:
                    self._left_rotate(parent)
            elif x == parent.left and parent == grandparent.left:  # Zig-Zig (left-left)
                self._right_rotate(grandparent)
                self._right_rotate(parent)
            elif x == parent.right and parent == grandparent.right:  # Zig-Zig (right-right)
                self._left_rotate(grandparent)
                self._left_rotate(parent)
            elif x == parent.right and parent == grandparent.left:  # Zig-Zag (left-right)
                self._left_rotate(parent)
                self._right_rotate(grandparent)
            else:  # Zig-Zag (right-left)
                self._right_rotate(parent)
                self._left_rotate(grandparent)

    def search(self, key):
        """
        Searches for a key in the tree.

        If the key is found, the corresponding node is splayed to the root.
        If the key is not found, the last accessed node (the would-be parent)
        is splayed to the root.

        Args:
            key: The integer key to search for.

        Returns:
            bool: True if the key is found, False otherwise.
        """
        node = self.root
        last_node = None
        while node:
            last_node = node
            if key == node.key:
                self._splay(node)
                return True
            elif key < node.key:
                node = node.left
            else:
                node = node.right

        # If key is not found, splay the last visited node (if any)
        if last_node:
            self._splay(last_node)
        return False

    def insert(self, key):
        """
        Inserts a key into the tree.

        If the key already exists, the existing node is splayed to the root.
        If the key is new, it is inserted and then the new node is splayed
        to the root.

        Args:
            key: The integer key to insert.
        """
        # Find position for new node, and check for duplicates
        node = self.root
        parent = None
        while node:
            parent = node
            if key == node.key:
                # Key already exists, splay this node and return
                self._splay(node)
                return
            elif key < node.key:
                node = node.left
            else:
                node = node.right

        # Insert the new node
        new_node = self.Node(key)
        new_node.parent = parent
        if not parent:
            self.root = new_node
        elif key < parent.key:
            parent.left = new_node
        else:
            parent.right = new_node

        # Splay the new node to the root
        self._splay(new_node)

    def delete(self, key):
        """
        Deletes a key from the tree.

        First, the node with the given key is splayed to the root. If it
        doesn't exist, the nearest node is splayed and the function returns.
        If it exists, it is removed, and its two subtrees are joined.

        Args:
            key: The integer key to delete.
        """
        # search() will splay the node (or its parent) to the root
        if not self.search(key):
            # Key was not in the tree, search already splayed the closest node
            return

        # At this point, the node to delete is the root because search(key) was True
        left_subtree = self.root.left
        right_subtree = self.root.right

        if not left_subtree:
            # If no left child, the right subtree becomes the new tree
            self.root = right_subtree
            if self.root:
                self.root.parent = None
        else:
            # Disconnect left subtree to work on it
            left_subtree.parent = None

            # Find the maximum element in the left subtree
            max_in_left = left_subtree
            while max_in_left.right:
                max_in_left = max_in_left.right

            # Splay this maximum element to the root of the left subtree.
            # After this, the new root of the left subtree (max_in_left)
            # will have no right child. We achieve this by temporarily
            # setting the left subtree as the main tree to reuse _splay.
            self.root = left_subtree
            self._splay(max_in_left)

            # self.root is now the root of the modified left subtree.
            # We can now attach the original right subtree.
            self.root.right = right_subtree
            if right_subtree:
                right_subtree.parent = self.root